src/main/cpp/src/decimal_utils.cu

```diff
@@ -520,6 +519,72 @@ inline __device__ int precision10(chunked256 value) {
     return -1;
 }
 
+__device__ chunked256 set_scale_and_round(chunked256 data, int old_scale, int new_scale) {
+    if (old_scale != new_scale) {
+        if (new_scale < old_scale) {
+            int raise = old_scale - new_scale;
+            int multiplier = pow_ten(raise).as_128_bits();
+            data = multiply(data, chunked256(multiplier));
+        } else {
+            int drop = new_scale - old_scale;
+            auto const diviser = pow_ten(drop).as_128_bits();
+            data = divide_and_round(data, chunked256(diviser).as_128_bits());
+        }
+    }
+    return data;
+}
+
+// Functor to add two DECIMAL128 columns with rounding and overflow detection.
+struct dec128_add: public thrust::unary_function<cudf::size_type, __int128_t> {
+  dec128_add(bool *overflows, cudf::mutable_column_view const &sum_view,
+                    cudf::column_view const &a_col, cudf::column_view const &b_col)
+      : overflows(overflows), a_data(a_col.data<__int128_t>()), b_data(b_col.data<__int128_t>()),
+        add_data(sum_view.data<__int128_t>()),
+        a_scale(a_col.type().scale()), b_scale(b_col.type().scale()),
+        sum_scale(sum_view.type().scale()) {}
+
+  __device__ __int128_t operator()(cudf::size_type const i) const {
+    chunked256 const a(a_data[i]);
+    chunked256 const b(b_data[i]);
+
+    chunked256 working_a = a;
+    chunked256 working_b = b;
+
+    int intermediate_scale = min(a_scale, b_scale);
+    if (a_scale != intermediate_scale) {
+    printf("converting a_scale \n");
+        working_a = set_scale_and_round(working_a, a_scale, intermediate_scale);
+    }
+    if (b_scale != intermediate_scale) {
+    printf("converting b_scale \n");
+        working_b = set_scale_and_round(working_b, b_scale, intermediate_scale);
+    }
+
+    chunked256 sum = working_a;
+    sum.add(working_b);
+
+    if (sum_scale != intermediate_scale) {
+        sum = set_scale_and_round(sum, intermediate_scale, sum_scale);
```

Comment:
nit: the indentation appears to be off in a few places. We have been using 2 spaces for indentation.